did/peer.go

```diff
@@ -0,0 +1,539 @@
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/TBD54566975/ssi-sdk/crypto"
+	"github.com/TBD54566975/ssi-sdk/util"
+	"github.com/pkg/errors"
+)
+
+type DIDPeer string
+type PurposeType string
+
+var availablePeerMethods = map[string]Resolver{
+	"0": PeerMethod0{},
+	"1": PeerMethod1{},
+	"2": PeerMethod2{},
+}
+
+// ANBF specified here:
+// https://identity.foundation/peer-did-method-spec/#method-specific-identifier
+const (
+	PeerMethodPrefix                = "peer"
+	DIDPrefix                       = "did"
+	PeerTransform                   = "z"
+	PeerEncNumBasis                 = Base58BTCMultiBase
+	PeerDIDRegex                    = `^did:peer:(([01](z)([1-9a-km-zA-HJ-NP-Z]{46,47}))|(2((\.[AEVID](z)([1-9a-km-zA-HJ-NP-Z]{46,47}))+(\.(S)[0-9a-zA-Z=]*)?)))$`
+	PeerKnownContext                = "https://w3id.org/did/v1"
+	PeerDIDCommMessagingAbbr string = "dm"
+	PeerDIDCommMessaging     string = "DIDCommMessaging"
+	Hash                            = "#"
+)
+
+// Checks if peer DID is valid
+// https://identity.foundation/peer-did-method-spec/index.html#recognizing-and-handling-peer-dids
+func isPeerDID(did string) bool {
+	r, err := regexp.Compile(PeerDIDRegex)
+	if err != nil {
+		panic(err)
+	}
+	return r.MatchString(did)
+}
+
+// Checks if the Peer DID is correctly
+// formatted
+func (d DIDPeer) IsValid() bool {
+	return isPeerDID(string(d))
+}
+
+func (d DIDPeer) Parse() (string, error) {
+
+	s, err := ParseDID(d, DIDPrefix+":"+PeerMethodPrefix+":")
+
+	if err != nil {
+		return "", err
+	}
+
+	method, err := d.GetMethodId()
+	if err != nil {
+		return "", err
+	}
+
+	var index int
+	switch method {
+	case "0":
+		index = 1
+	case "1":
+		return "", errors.Wrap(util.NotImplementedError, "parsing method 1")
+	case "2":
+		index = 2
+	}
+	return s[index:], nil
+}
+
+// Method 0: inception key without doc
+// https://identity.foundation/peer-did-method-spec/index.html#generation-method
+// The DID doc offers no endpoint. This makes the DID functionally equivalent to
+// a did:key value For example,
+// did:key:z6MkpTHR8VNsBxYAAWHut2Geadd9jSwuBV8xRoAnwWsdvktH is equivalent to
+// did:peer:0z6MkpTHR8VNsBxYAAWHut2Geadd9jSwuBV8xRoAnwWsdvktH
+type PeerMethod0 struct {
+	kt crypto.KeyType
+}
+
+func buildDIDPeerFromEncoded(method int, encoded string) DIDPeer {
+	return DIDPeer(fmt.Sprintf("%s:%s:%d%s", DIDPrefix, PeerMethodPrefix, method, encoded))
+}
+
+type byValue struct {
+	Key       string `json:"key"`
+	Signature string `json:"sig"`
+}
+
+//https://identity.foundation/peer-did-method-spec/#backing-storage
+type PeerDelta struct {
+	Change string    `json:"change"` // <base64url encoding of a change fragment>,
+	By     []byValue `json:"by"`     //  [ {"key": <id of key>, "sig": <signature value>} ... ],
+	When   int64     `json:"when"`   //<ISO8601/RFC3339 UTC timestamp with at least second precision>
+}
+
+func (a DIDPeer) Delta(b DIDPeer) (PeerDelta, error) {
+	return PeerDelta{}, errors.Wrap(util.NotImplementedError, "peer:did delta")
```

Comment:
what do you think about returning `(*PeerDelta, error)`? I usually like pointer returns to simply return `nil, err` instead of instantiating an empty instance of the struct